### ISSUE TITLE
SparkApplication submit fails: `Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Operation: [create]  for kind: [Pod]  with name: [null]  in namespace: [my-ns]  failed.` when running with webhook `enable: true`

### ISSUE THREAD

**Issue by JunaidChaudry** [GoogleCloudPlatform/spark-on-k8s-operator#1708]
Hi, 

I have the latest spark-operator installed on kubernetes version 1.23, but My `SparkApplication` is failing to submit with the following error in the operator logs: 

``` 
Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: Operation: [create]  for kind: [Pod]  with name: [null]  in namespace: [spark-operator]  failed.
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
        at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:72)
        at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:349)
        at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:84)
        at org.apache.spark.deploy.k8s.submit.Client.run(KubernetesClientApplication.scala:139)
        at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$3(KubernetesClientApplication.scala:213)
        at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.$anonfun$run$3$adapted(KubernetesClientApplication.scala:207)
        at org.apache.spark.util.Utils$.tryWithResource(Utils.scala:2611)
        at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.run(KubernetesClientApplication.scala:207)
        at org.apache.spark.deploy.k8s.submit.KubernetesClientApplication.start(KubernetesClientApplication.scala:179)
        at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:951)
        at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
        at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
        at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
        at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1030)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1039)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.net.SocketTimeoutException: timeout
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.java:672)
        at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.java:680)
        at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.java:153)
        at okhttp3.internal.http2.Http2Codec.readResponseHeaders(Http2Codec.java:131)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at io.fabric8.kubernetes.client.utils.BackwardsCompatibilityInterceptor.intercept(BackwardsCompatibilityInterceptor.java:135)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at io.fabric8.kubernetes.client.utils.OIDCTokenRefreshInterceptor.intercept(OIDCTokenRefreshInterceptor.java:41)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at io.fabric8.kubernetes.client.utils.ImpersonatorInterceptor.intercept(ImpersonatorInterceptor.java:68)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at io.fabric8.kubernetes.client.utils.HttpClientUtils.lambda$createHttpClient$3(HttpClientUtils.java:151)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
        at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257)
        at okhttp3.RealCall.execute(RealCall.java:93)
        at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:490)
        at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
        at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:252)
        at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:879)
        at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:341)
        ... 14 more
23/03/27 16:15:03 INFO ShutdownHookManager: Shutdown hook called
23/03/27 16:15:03 INFO ShutdownHookManager: Deleting directory /tmp/spark-f18eea0e-6437-4444-b2fd-e429aedbf6b6

```

Submitting ANY `SparkApplication` has the same result and the same error as above. The serviceaccount being used with `SparkApplication` has the following roles and RoleBinding and role: 
```yaml
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: my-serviceaccount
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: spark-role
rules:
- apiGroups: [""]
  resources: ["pods", "configmaps"]
  verbs: ["*"]
- apiGroups: [""]
  resources: ["services"]
  verbs: ["*"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: spark-role-binding
subjects:
- kind: ServiceAccount
  name: my-serviceaccount
roleRef:
  kind: Role
  name: spark-role
  apiGroup: rbac.authorization.k8s.io
```
My `SparkApplication` file is: 

```yaml
apiVersion: "sparkoperator.k8s.io/v1beta2"
kind: SparkApplication
metadata:
  name: patient
  namespace: spark-operator
spec:
  sparkConf:
    "spark.kubernetes.local.dirs.tmpfs": "true"
  type: Python
  mode: cluster
  mainClass: org.apache.spark.deploy.PythonRunner
  pythonVersion: "3"
  image: "gcr.io/spark-operator/spark:v3.1.1"
  imagePullPolicy: Always
  mainApplicationFile: "local:///app/asdf.py"
  sparkVersion: "3.0.0"
  restartPolicy:
    type: Never

  dynamicAllocation:
    enabled: true
    initialExecutors: 1
    minExecutors: 1
    maxExecutors: 50
  driver:
    cores: 1
    coreLimit: "1200m"
    memory: "512m"     
    labels:
      version: 3.0.0
    serviceAccount: my-serviceaccount
    env:
    - name: HTTP2_DISABLE # https://github.com/fabric8io/kubernetes-client/issues/2212#issuecomment-628551315
      value: "true"   
    - name: KUBERNETES_TLS_VERSIONS
      value: TLSv1.2,TLSv1.3
  executor:
    cores: 1
    instances: 1
    memory: "512m"    
    labels:
      version: 3.0.0
    env:
    - name: HTTP2_DISABLE # https://github.com/fabric8io/kubernetes-client/issues/2212#issuecomment-628551315
      value: "true"  
    - name: KUBERNETES_TLS_VERSIONS
      value: TLSv1.2,TLSv1.3
```

**System Information**: 
Kubernetes type: EKS
Kubernetes Version: `Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.16-eks-48e63af", GitCommit:"e6332a8a3feb9e0fe3db851878f88cb73d49dd7a", GitTreeState:"clean", BuildDate:"2023-01-24T19:18:15Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}`
Helm Chart Version:  `1.1.27`
Spark-Operator Image:Version: `spark-operator:v1beta2-1.3.8-3.1.1`

**To reproduce:**

1. Install latest spark-on-k8s helm chart with webhook enabled
2. Apply the required role and rolebindings
3. Try submitting ANY sparkApplication

**Comment by ricardopalazzio** [comment]
Hi [JunaidChaudry](https://github.com/JunaidChaudry),  I solved allowing all traffic in the vpc's ip range, in the node's security-group (EKS node shared security group)

**Comment by JunaidChaudry** [comment]
@ricardopalazzio I forgot to mention, if I install kubectl in the operator pod, and try submitting a resource from within the pod using `kubectl apply`, that works totally fine. So I doubt it has to do with connections from the operator to the cluster

**Comment by JunaidChaudry** [comment]
I also just double checked... The security group is already allowing inbound and outbound traffic to all ports and protocols within the VPC

**Comment by JunaidChaudry** [comment]
Does anyone have any other recommendations please? I still haven't been able to find a solution for this

**Comment by JunaidChaudry** [comment]
I have also confirmed that the issue only happens when webhook is enabled on spark-operator

**Comment by alita1991** [comment]
I have a similar issue, anyone has an idea how to debug this? I just receive timeout but have no clue where to actually look, the error message does not help at all

**Comment by JunaidChaudry** [comment]
I was able to find the workaround for it. Apparently there is a spark conf that can be used to increase the timeout limits https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/1357#issuecomment-1218396290 . Setting the following `sparkConf` properties seemed to unblock me: 

```yaml
sparkConf: 
  "spark.kubernetes.local.dirs.tmpfs": "true"
  "spark.kubernetes.submission.connectionTimeout": "60000000"
  "spark.kubernetes.submission.requestTimeout": "60000000" 
  "spark.kubernetes.driver.connectionTimeout": "60000000" 
  "spark.kubernetes.driver.requestTimeout": "60000000"
```

**Comment by JunaidChaudry** [comment]
However, now I have a different issue where any of the `driver` and `executor` properties which require webhook, are not getting passed in to the driver and executor pods. For example [`tolerations`](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/user-guide.md#using-tolerations) that I have added to the driver and executor pods are not part of the corresponding pods. The same goes for [`volumeMounts`](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/user-guide.md#mounting-volumes)

**Comment by JunaidChaudry** [comment]
@nikoshet  That seemed to have worked very nicely. Thank you so much! 

If you don't mind me asking, how did you figure out that the port needed to be changed?

**Comment by nikoshet** [comment]
@JunaidChaudry I read [this](https://googlecloudplatform.github.io/spark-on-k8s-operator/docs/quick-start-guide.html#mutating-admission-webhooks-on-a-private-gke-cluster), and although my cluster is not deployed on GKE but it is in a private subnet, I guessed that the Kubernetes master has the same behavior and only allows secure connections on port 443. 😄

**Comment by JunaidChaudry** [comment]
> @JunaidChaudry I read [this](https://googlecloudplatform.github.io/spark-on-k8s-operator/docs/quick-start-guide.html#mutating-admission-webhooks-on-a-private-gke-cluster), and although my cluster is not deployed on GKE but it is in a private subnet, I guessed that the Kubernetes master has the same behavior and only allows secure connections on port 443. 😄

My set up is also the same (aws with private subnets). Its definitely a pretty weird fix however. It used to work until the beginning of 2023 without needing this change, and then suddenly stopped working once I updated my nodes to the latest ami. Much Thanks however. It fixed the issue for me too ^_^
